internal/selfupdate/selfupdate.go

```diff
@@ -0,0 +1,341 @@
+
+type HTTPClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+type SelfUpdater struct {
+	version      string
+	noSelfUpdate bool
+	releaseTrack string
+	httpClient   HTTPClient
+}
+
+type Release struct {
+	Version     string `json:"version"`
+	Description string `json:"description"`
+	AssetURL    string `json:"assetURL"`
+	SHA256      string `json:"sha256"`
+}
+
+const (
+	appName = "Zen"
+)
+
+func NewSelfUpdater(httpClient HTTPClient) (*SelfUpdater, error) {
+	if httpClient == nil {
+		return nil, errors.New("httpClient is nil")
+	}
+
+	u := SelfUpdater{
+		version:      Version,
+		releaseTrack: releaseTrack,
+		httpClient:   httpClient,
+	}
+	switch noSelfUpdate {
+	case "true":
+		u.noSelfUpdate = true
+	case "false":
+	default:
+		return nil, fmt.Errorf("invalid noSelfUpdate value: %s", noSelfUpdate)
+	}
+
+	return &u, nil
+}
+
+func (su *SelfUpdater) checkForUpdates() (*Release, error) {
+	log.Println("checking for updates")
+	if su.noSelfUpdate {
+		log.Println("noSelfUpdate=true, self-update disabled")
+		return nil, nil
+	}
+	if su.version == "development" {
+		log.Println("version=development, self-update disabled")
+		return nil, nil
+	}
+
+	url := fmt.Sprintf("%s/%s/%s/%s/manifest.json", manifestsBaseURL, su.releaseTrack, runtime.GOOS, runtime.GOARCH)
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return nil, fmt.Errorf("create request: %w", err)
+	}
+
+	req.Header.Set("Accept", "application/json")
+	req.Header.Set("User-Agent", "zen-desktop")
+
+	res, err := su.httpClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("do request: %w", err)
+	}
+
+	defer res.Body.Close()
+
+	var rel Release
+	if err := json.NewDecoder(res.Body).Decode(&rel); err != nil {
+		return nil, fmt.Errorf("decode response: %w", err)
+	}
+
+	return &rel, nil
+}
+
+// isNewer compares the current version with the version passed as an argument and returns true if the argument is newer.
+//
+// It assumes that both versions are in the format "v<major>.<minor>.<patch>" and returns an error if they are not.
+func (su *SelfUpdater) isNewer(version string) (bool, error) {
+	var currentMajor, currentMinor, currentPatch, newMajor, newMinor, newPatch int
+	if _, err := fmt.Sscanf(su.version, "v%d.%d.%d", &currentMajor, &currentMinor, &currentPatch); err != nil {
+		return false, fmt.Errorf("parse current version (%s): %w", su.version, err)
+	}
+	if _, err := fmt.Sscanf(version, "v%d.%d.%d", &newMajor, &newMinor, &newPatch); err != nil {
+		return false, fmt.Errorf("parse new version (%s): %w", version, err)
+	}
+
+	if newMajor > currentMajor {
+		return true, nil
+	}
+	if newMajor == currentMajor && newMinor > currentMinor {
+		return true, nil
+	}
+	if newMajor == currentMajor && newMinor == currentMinor && newPatch > currentPatch {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (su *SelfUpdater) ApplyUpdate(ctx context.Context) error {
```

Comment:
The function seems a bit heavy and difficult to read at a glance. How about breaking it down into smaller, more manageable functions?